File: LinkedList/reverseList.cpp
/*
* #: 206
* Title: Reversed Linked List
* Description:
* ------
* Reverse a singly linked list.
* ------
* Time: O(n)
* Space: O(1)
* Difficulty:  Easy
*/
#include <iostream>
#include <vector>

// Definition of singly linked list
struct ListNode{
    int val;
    ListNode* next;
    ListNode(int x) : val(x), next(nullptr) {}
};

class Solution{
public:
    ListNode* reverseList(ListNode* head){
        ListNode* prev{};
        ListNode* mid{head};
        ListNode* next;

        while(mid){
            next = mid->next;
            mid->next = prev;
            prev = mid;
            mid = next;
        }
        return prev;
    }
};

int main(void)
{
    std::vector<int> L1 = {1,3,5,7,8,9};
    ListNode n1 = ListNode(L1[0]);

    ListNode* curt = &n1;

    for (unsigned int i=1; i < L1.size(); ++i){
        curt->next = new ListNode(L1[i]);
        curt = curt->next;
    }
    std::cout << "before reverse:\n";
    curt =&n1;
    while (curt){
        std::cout << curt->val << ", ";
        curt = curt->next;
    }
    Solution sol = Solution();
    ListNode* result = sol.reverseList(&n1);
    std::cout << "\n after reverse: \n";
    curt = result;
    while (curt){
        std::cout << curt->val << ", ";
        curt = curt->next;
    }
    std::cout << "\n";
    return 0;
}

File: LinkedList/reverseList.py
"""
#: 206
Title: Reverse Linked List
Description:
------
Reverse a singly linked list
------
Time: O(n)
Space: O(1)
Difficulty: Easy
"""
# Definition for singly-linked list.
class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution(object):
    def reverseList(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        prev = None
        next = None
        current = head
        while current:
            next = current.next
            current.next = prev
            prev = current
            current = next
        return prev


if __name__ == '__main__':
    L1 = [1,3,5,8,7,9,12]
    n1 = ListNode(L1[0])
    curt = n1
    for i in L1[1:]:
        curt.next = ListNode(i)
        curt = curt.next
    sol = Solution()
    print 'before reverse linked list:'
    curt = n1
    tmp = []
    while(curt):
        tmp.append(curt.val)
        curt = curt.next
    print tmp

    res = sol.reverseList(n1)
    print 'after reverse:'
    curt = res
    tmp2 = []
    while(curt):
        tmp2.append(curt.val)
        curt = curt.next
    print tmp2


